### DISCUSSION TITLE
Compiling .py files with Cython?

### DISCUSSION THREAD

**Discussion by madig** [mesonbuild/meson#11578] [Q&A] [answered]
Hi!
I'm trying my hand at converting [lxml](https://github.com/lxml/lxml) to use meson-python. The project cythonizes two .pyx files but also a couple of .py files. Just adding them like so:

```
py.extension_module(
    'lxml',
    'src/lxml/_elementpath.py',
    'src/lxml/builder.py',
    'src/lxml/etree.pyx',
    'src/lxml/html/clean.py',
    'src/lxml/html/diff.py',
    'src/lxml/objectify.pyx',
    'src/lxml/sax.py',
    dependencies: [dep_py, dep_libxml2, dep_libxslt],
    include_directories: ['src/lxml/includes'],
    install: true
)
```

makes meson complain that it doesn't know what to do with .py files. How do I tell it to do the same thing as for .pyx files?

**Comment by eli-schwartz** [comment, marked as answer]
You can do:

```
cython = find_program('cython')

c_file = custom_target(
    input: 'sax.py',
    output: 'sax.c',
    command: [cython, '-3', '@INPUT@', '-o', '@OUTPUT@']
)

py.extension_module('sax', c_file, install: true)
```

Because each compiled extension needs to be listed separately as a different "py.extension_module", and pure-python that has been cythonized is one extension module per .py file with no dependencies.

Note that it *is* on my TODO list to add some method to handle automatically cythonizing .py files for you, as I'd like this to be more DRY.

**Reply by madig** [reply]
Thanks for the pointer :)
